gtsam/geometry/Pose3.cpp

```diff
@@ -63,6 +63,81 @@ Matrix6 Pose3::AdjointMap() const {
   return adj;
 }
 
+/* ************************************************************************* */
+// Calculate AdjointMap applied to xi_b, with Jacobians
+Vector6 Pose3::Adjoint(const Vector6& xi_b, OptionalJacobian<6, 6> H_this,
+                       OptionalJacobian<6, 6> H_xib) const {
+  //  Ad   *   xi   =   [   R   0    .  [w
+  //                      [t]R  R ]      v]
+  // Declarations and aliases
+  Matrix3 Rw_H_R, Rw_H_w, Rv_H_R, Rv_H_v, pRw_H_t, pRw_H_Rw;
+  Vector6 result;
+  auto Rw = result.head<3>();
+  const Vector3 &w = xi_b.head<3>(), &v = xi_b.tail<3>();
+
+  // Calculations
+  Rw = R_.rotate(w, H_this ? &Rw_H_R : nullptr, H_xib ? &Rw_H_w : nullptr);
+  const Vector3 Rv =
+      R_.rotate(v, H_this ? &Rv_H_R : nullptr, H_xib ? &Rv_H_v : nullptr);
+  const Vector3 pRw = cross(t_, Rw, H_this ? &pRw_H_t : nullptr,
+                            (H_this || H_xib) ? &pRw_H_Rw : nullptr);
+  result.tail<3>() = pRw + Rv;
+
+  // Jacobians
+  if (H_this) {
+    pRw_H_t = Rw_H_R;  // This is needed to pass the unit tests for some reason
```

Comment:
We should figure this out before landing this.

Reply:
Uhm, yes. Need some math. In math.pdf?

Reply:
It seems like the way you're doing the `cross` operation is making the jacobian be incorrect. The jacobian should be `d (p_cross * R * w) / dp`.

By the chain rule for matrix-matrix-products [Matrix Cookbook Eq 37](https://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf), we get `R * dp_cross/dp * w + p_cross * dR/dp * w = R * dp_cross/dp * w`.

We have `p_cross` as:
```
p_cross = 
 0 -z  y
 z  0 -x
-y  x  0
```

If we differentiate the skew-symmetric matrix with each element, we get:
```
dp_cross / dx = 
0 0  0
0 0 -1
0 1  0

dp_cross / dy = 
0  0 1
0  0 0
-1 1 0

dp_cross / dz = 
0 -1 0
1  0 0
0  0 0
```

Just for `dx` we get
```
R11 R12 R13 * 0 0  0 * wx
R21 R22 R23   0 0 -1   wy
R31 R32 R33   0 1  0   wz

=
0  R13  -R12   wx
0  R23  -R22 * wy
0  R33  -R32   wz

=
R13*wy - R12*wz
R23*wy - R22*wz
R33*wy - R32*wz
```

which is the same as `R * (-w_cross)`.
```
R11 R12 R13 *  0   wz  -wy
R21 R22 R23   -wz  0    wx
R31 R32 R33    wy -wx   0

=
(-R12*wz + R13*wy)
(-R22*wz + R23*wy)
(-R32*wz + R33*wy)
```

This is why using `Rw_H_R` works, since it is just `R * (-w_cross)`.


#### Possible Solutions

1. Compute the jacobian `pRw_H_t` separately as `R * (-w_cross)`.
2. Implement the cross product for `Rot3` with a `Point3` along with the appropriate jacobians.

I personally prefer (1) since it is easier, though we will have to make a note about the reason for it (which can be just the comments from here).

Reply:
Thanks for this analysis!

But I have a couple questions:

1) >By the chain rule for matrix-matrix-products Matrix Cookbook Eq 37,...
    
    did you mean product rule?  Eq 37 states ![\partial{ XY } = (\partial{X})Y + X(\partial{Y})](https://user-images.githubusercontent.com/25356161/137571902-9876f2a5-0cce-4923-b34d-a368f9fdc65c.png)
2) >... we get `R * dp_cross/dp * w + p_cross * dR/dp * w = R * dp_cross/dp * w`.  
    
    I'm not sure I understand this - Eq 37 doesn't switch the order of X and Y so I think it should be (using [] to denote skew symmetric matrix):
    ![image](https://user-images.githubusercontent.com/25356161/137572084-ae251b44-25ae-4980-9c31-436f0d38cff3.png)
    instead of yours (typeset to be easier to read):
    ![image](https://user-images.githubusercontent.com/25356161/137572206-68956ec3-c346-4039-a109-4cf39c289173.png)
    And then I think if you switch the order of \partial{[t]}R then I think we can simplify (abusing the skew-symmetric notation to apply to matrices to indicate "column-wise" and forming a tensor):
    ![image](https://user-images.githubusercontent.com/25356161/137572276-ca326be4-6f4a-4886-8ffb-433908183024.png)
    (this is I think the same identity that you maybe noticed: `R[I]w = -RI[w]` where I is 3x3 identity)
    And this expression is the same expression in `Point3::cross`: `if (H1) *H1 << skewSymmetric(-q.x(), -q.y(), -q.z());`
3) `Rw` is an ordinary 3-vector so why wouldn't `t x (Rw)` be able to use the cross product Jacobians?  I think the explanation you give would give different results if you tried to do `\partial (tR) x w` vs `\partial t x (Rw)`, which of course these should give the same result

I suspect the actual problem with this jacobian is that `Pose3::Local` is NOT just `Vector6() << R_.Local(), t_.Local()`.  Instead, it's applying some stuff to `t_` (see `Pose3::Logmap`).  But the problem with this hypothesis is that turning off `GTSAM_POSE3_EXPMAP` does not resolve this, so I'm still working on figuring out what's going on.

PS: the same issue occurs in [`AdjointTranspose` calculating `H_this`](https://github.com/borglab/gtsam/blob/2de623befccc6cce0b064737053e3ca4d9a72389/gtsam/geometry/Pose3.cpp#L132-L136) and I'm not sure if the same reasoning applies to this one because this code does the cross product before doing the rotation.

Reply:
Following up on my "edit":

> I forgot that T.localCoordinates is between-ing T then applying ChartAtOrigin so this is where a multiplication by R gets introduced.

`-[Rw] * R = -(Rw) x R = R[w]`
That last part can be intuited by splitting up ![image](https://user-images.githubusercontent.com/25356161/137573577-41a8637b-4822-45ba-9507-befea3614ba7.png) where `R_i` denotes the i'th column of R.  Then geometrically you can imagine that `-w_i(R_i x R)` will generate something similar to `R[w]` since `R_i x R_j` will be 0 if i==j and `+/-R_k` otherwise (where k =/= i =/= j).  Carefully working through each example we can confirm that indeed it is exactly `R[w]`.

This correction (multiply by R) works for both `Adjoint` and `AdjointTranspose` so I think this was indeed the issue.

In summary, for some arbitrary `X`, `[dX / dT] = [dX/dR ; dX/dt * R]` where the pose `T` has rotation `R` and translation `t`, where the extra `R` is the part I was missing.  Furthermore, some algebraic manipulation can simplify the expressions.

I'll write this up and simplify.

Reply:
You're right about the ordering for the product rule (I don't know why I called it the chain rule, maybe because it is a chain of operations?). This [stackexchange answer](https://math.stackexchange.com/a/366931/134918) was what gave me the equation and I just looked up the equation number.

Reply:
> the issue is exactly that the multiplication by R is not happening since the cross product is not aware of that operation having happened.

It doesn't need to be aware of that multiplication?  `Rw` is an ordinary 3-vector so I don't really see why it matters where `Rw` came from when calculating jacobian w.r.t. `t` since `Rw` is independent of `t` (but if there were a dependence, then we would use product rule and include a term `+ pRw_H_Rw * Rw_H_t` and we still wouldn't need to do this "manual" matrix math work).  Basically, I don't think we should ever need to drop into the matrix math as long as we just follow chain rule.  I think this is most clear in commit 33e16aa, but then in f2ac90d I made the tradeoff to be more efficient at the cost of readability by applying the simplification that `pRw_H_t * t_H_xi = pRw_H_t * R = Rw_H_R` by some matrix magic.

> I think we came to the same conclusion but by different means. I'll spend more time understanding your comment and see what is the key difference.

I don't think so - I don't think the issue is related to cross product at all, but also feel free to check out the math.pdf updates I made in Sec. 13.2.  The key sentence for this discussion being "Applying chain rule for the translation ...".

### Simplest Explanation
`H_this` does not equal `[H_R , H_t]`, it actually equals `[H_R , H_t] * [R_H_this ; t_H_this]`.  Basically, I assumed that the "vector representation for Pose3" (aka Local) was just R stacked on top of t (which is false) and therefore I forgot about the partial derivatives w.r.t. R, t.

If you look at [`translation(Hself)` and `rotation(Hself)`](https://github.com/borglab/gtsam/blob/f2ac90d5b9682f4b78d8ffe99be378ed32df11c5/gtsam/geometry/Pose3.cpp#L344-L357) (reproduced below), you will see that actually `translation(Hself)` is _not_ `[0 I]`, it is `[0 R]`.

https://github.com/borglab/gtsam/blob/f2ac90d5b9682f4b78d8ffe99be378ed32df11c5/gtsam/geometry/Pose3.cpp#L344-L357

### Explanation for why `translation(Hself)` is `[0 R]` (optional read if you're curious, but not necessary to understand)
Recall that when we do derivative w.r.t. a manifold object, what we actually mean is something along the lines of:
![image](https://user-images.githubusercontent.com/25356161/137594129-614b2dc4-9ed6-4b6e-ba37-5608b55abe5f.png)
Where `f(a)` is the function we're trying to differentiate (`this.Adjoint() * xi_b` in this case) and `R_a(xi)` denotes the retraction at pose `a` using local coordinates `xi`.  This is **NOT** mathematically precise, just what I thought was a more intuitive version of the formal definition in Section 6.2 of math.pdf:
![image](https://user-images.githubusercontent.com/25356161/137594063-c910ef2b-0ac4-4a05-8158-047204e0ff6d.png)

The key being that it's the retraction *at `a`* (aka `this`).  This is significant because the jacobian we calculated was w.r.t. the translation in the world frame, but actually we want the translation in the frame of `this`.   We would be able to see this either in `retract` or `localCoordinates` of `Lie.h`:
https://github.com/borglab/gtsam/blob/9798bfa8151788b31e225be9acce0ca4c4d30b10/gtsam/base/Lie.h#L140-L161

Notice how there is a `compose` and `between` in `retract` / `localCoordinates` respectively, so although the ChartAtOrigin does indeed just stack `xi = [R t]`, we have to transform this into the chart at `a` by rotating the translation by `R.transpose()`.  The transpose goes away because in the jacobian the translation is across columns so instead of `R.transpose * xi` it's more like `xi.transpose() * R` (colloquially - it's hard to write in text but I'm sure you can figure this out).

Reply:
This is probably the clearest code, from the state of the code at 33e16aa

https://github.com/borglab/gtsam/blob/33e16aa7d2ba47239ec59439da997238d642b8bd/gtsam/geometry/Pose3.cpp#L67-L99